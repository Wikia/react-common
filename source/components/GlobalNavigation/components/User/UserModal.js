import React, { useEffect, useContext } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { useTranslation } from 'react-i18next';

import Avatar from '../../../Avatar';

import NotificationsContext from '../../context/NotificationContext';
import I18nNamespaceContext from '../../context/I18nNamespaceContext';

import LinkLogOut from '../Link/LinkLogOut';
import NotificationsListContainer from '../Notifications/ListContainer';

const UserModal = ({ isOpen, user, track, header }) => {
    const { loadFirstPage } = useContext(NotificationsContext);
    const [t] = useTranslation(useContext(I18nNamespaceContext));
    const viewProfileLink = user.items.filter(item => item['tracking-label'] === 'account.profile')[0].href;
    const signOutModel = user.items.filter(item => item.type === 'link-logout')[0];
    const classes = classNames('wds-global-navigation__modal', 'wds-user-modal', { 'wds-is-hidden': !isOpen });

    useEffect(() => {
        if (isOpen) {
            loadFirstPage();
        }
    }, [isOpen, loadFirstPage]);

    return (
        <div className={classes}>
            <div className="wds-user-modal__info">
                <Avatar src={user.avatar_url} href={viewProfileLink} data-tracking-label="open-user-profile" />
                <p className="wds-user-modal__username">
                    {user.username}
                </p>
            </div>
            <div className="wds-user-modal__notifications-wrapper">
                <NotificationsListContainer track={track} header={t(header)} />
            </div>
            <LinkLogOut link={signOutModel} className="wds-user-modal__sign-out-container" />
        </div>
    );
};

UserModal.propTypes = {
    header: PropTypes.string.isRequired,
    isOpen: PropTypes.bool.isRequired,
    track: PropTypes.func.isRequired,
    user: PropTypes.shape().isRequired,
};

export default UserModal;
