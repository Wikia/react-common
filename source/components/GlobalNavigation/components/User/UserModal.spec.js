import React, { useEffect, useContext } from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import merge from 'lodash/merge';

import UserModal from './UserModal';

const defaultProps = {
    user: {
        avatar_url: 'https://fandomdesignsystem.com/images/FANDOM-Avatar.jpg',
        username: 'Ryba777',
        'tracking-label': 'account',
        items: [
            {
                type: 'link-text',
                title: {
                    type: 'translatable-text',
                    key: 'global-navigation-user-view-profile',
                },
                href: '//starwars.wikia.com/wiki/User:Ryba777',
                'tracking-label': 'account.profile',
            },
            {
                type: 'link-text',
                title: {
                    type: 'translatable-text',
                    key: 'global-navigation-user-my-talk',
                },
                href: '//starwars.wikia.com/wiki/User_talk:Ryba777',
                'tracking-label': 'account.talk',
            },
            {
                type: 'link-text',
                title: {
                    type: 'translatable-text',
                    key: 'global-navigation-user-my-preferences',
                },
                href: '//starwars.wikia.com/wiki/Special:Preferences',
                'tracking-label': 'account.preferences',
            },
            {
                type: 'link-text',
                title: {
                    type: 'translatable-text',
                    key: 'global-navigation-user-help',
                },
                href: '//community.wikia.com/wiki/Help:Contents',
                'tracking-label': 'account.help',
            },
            {
                type: 'link-logout',
                href: 'https://www.wikia.com/logout',
                title: {
                    type: 'translatable-text',
                    key: 'global-navigation-user-sign-out',
                },
                'param-name': 'returnto',
                'tracking-label': 'account.sign-out',
            },
        ],
    },
    track: () => null,
    isOpen: false,
};

const mockDefaultContext = { loadFirstPage: () => null };

function renderComponent(props) {
    const renderer = new ShallowRenderer();
    const computedProps = merge({}, defaultProps, props);

    renderer.render(<UserModal {...computedProps} />);

    return renderer.getRenderOutput();
}

// TODO: remove mocks when enzyme/react-test-renderer will fully support hooks
jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useContext: jest.fn(() => mockDefaultContext),
    useEffect: jest.fn(),
}));

/* eslint-disable-next-line jest/no-hooks */
beforeEach(() => {
    jest.resetModules();
});

test('UserModal renders correctly with default params', () => {
    expect(renderComponent()).toMatchSnapshot();
});

test('UserModal renders correctly with isOpen set as true', () => {
    expect(renderComponent({ isOpen: true })).toMatchSnapshot();
});

test('UserModal does not call loadFirstPage via useEffect when isOpen is false', () => {
    const loadFirstPageMock = jest.fn();
    useContext.mockImplementation(() => ({ loadFirstPage: loadFirstPageMock }));
    useEffect.mockImplementation(callback => callback());

    renderComponent();

    expect(loadFirstPageMock).not.toBeCalled();
});

test('UserModal calls loadFirstPage via useEffect when isOpen is true', () => {
    const loadFirstPageMock = jest.fn();
    useContext.mockImplementation(() => ({ loadFirstPage: loadFirstPageMock }));
    useEffect.mockImplementation(callback => callback());

    renderComponent({ isOpen: true });

    expect(loadFirstPageMock).toBeCalledWith();
});
