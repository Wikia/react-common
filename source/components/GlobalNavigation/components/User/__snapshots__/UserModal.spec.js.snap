// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserModal renders correctly with default params 1`] = `
<div
  className="wds-global-navigation__modal wds-user-modal wds-is-hidden"
>
  <div
    className="wds-user-modal__info"
  >
    <Avatar
      alt="User avatar"
      data-tracking-label="open-user-profile"
      href="//starwars.wikia.com/wiki/User:Ryba777"
      src="https://fandomdesignsystem.com/images/FANDOM-Avatar.jpg"
    />
    <p
      className="wds-user-modal__username"
    >
      Ryba777
    </p>
  </div>
  <div
    className="wds-user-modal__notifications-wrapper"
  >
    <ListContainer
      header="some-label"
      track={[Function]}
    />
  </div>
  <ComponentWrapper
    className="wds-user-modal__sign-out-container"
    link={
      Object {
        "href": "https://www.wikia.com/logout",
        "param-name": "returnto",
        "title": Object {
          "key": "global-navigation-user-sign-out",
          "type": "translatable-text",
        },
        "tracking-label": "account.sign-out",
        "type": "link-logout",
      }
    }
  />
</div>
`;

exports[`UserModal renders correctly with isOpen set as true 1`] = `
<div
  className="wds-global-navigation__modal wds-user-modal"
>
  <div
    className="wds-user-modal__info"
  >
    <Avatar
      alt="User avatar"
      data-tracking-label="open-user-profile"
      href="//starwars.wikia.com/wiki/User:Ryba777"
      src="https://fandomdesignsystem.com/images/FANDOM-Avatar.jpg"
    />
    <p
      className="wds-user-modal__username"
    >
      Ryba777
    </p>
  </div>
  <div
    className="wds-user-modal__notifications-wrapper"
  >
    <ListContainer
      header="some-label"
      track={[Function]}
    />
  </div>
  <ComponentWrapper
    className="wds-user-modal__sign-out-container"
    link={
      Object {
        "href": "https://www.wikia.com/logout",
        "param-name": "returnto",
        "title": Object {
          "key": "global-navigation-user-sign-out",
          "type": "translatable-text",
        },
        "tracking-label": "account.sign-out",
        "type": "link-logout",
      }
    }
  />
</div>
`;
